From 79d8235aa9b6036cf5840f52e5123f1cb6b248db Mon Sep 17 00:00:00 2001
From: sreichholf <stephan@reichholf.net>
Date: Tue, 26 Jan 2016 10:33:15 +0100
Subject: [PATCH] - start working on encoder based streaming

Conflicts:
	app/res/layout/profile_edit.xml
	app/res/values/strings.xml
	app/src/net/reichholf/dreamdroid/fragment/ProfileEditFragment.java
---
 app/res/layout/profile_edit.xml                    | 489 ++++++++++++---------
 app/res/values/strings.xml                         |   5 +
 app/src/net/reichholf/dreamdroid/Profile.java      | 152 ++++++-
 .../dreamdroid/fragment/ProfileEditFragment.java   |  71 ++-
 .../dreamdroid/helpers/SimpleHttpClient.java       |  27 ++
 .../dreamdroid/intents/IntentFactory.java          |   8 +-
 6 files changed, 522 insertions(+), 230 deletions(-)

diff --git a/app/res/layout/profile_edit.xml b/app/res/layout/profile_edit.xml
index e5f8012..03cbf93 100644
--- a/app/res/layout/profile_edit.xml
+++ b/app/res/layout/profile_edit.xml
@@ -1,279 +1,350 @@
 <?xml version="1.0" encoding="utf-8"?>
+<RelativeLayout xmlns:android="http://schemas.android.com/apk/res/android"
+                android:layout_width="fill_parent"
+                android:layout_height="fill_parent">
 
-<android.support.v4.widget.NestedScrollView
-    xmlns:android="http://schemas.android.com/apk/res/android"
-    android:layout_width="fill_parent"
-    android:layout_height="fill_parent"
-    android:paddingBottom="@dimen/content_vert_padding"
-    android:paddingLeft="@dimen/content_horz_padding"
-    android:paddingRight="@dimen/content_horz_padding"
-    android:paddingTop="@dimen/content_vert_padding">
-
-    <LinearLayout
+    <android.support.v4.widget.NestedScrollView
         android:layout_width="fill_parent"
-        android:layout_height="wrap_content"
-        android:orientation="vertical">
-
-        <EditText
-            android:id="@+id/EditTextProfile"
-            android:layout_width="fill_parent"
-            android:layout_height="0dip"
-            android:layout_marginBottom="4dp"
-            android:layout_weight="1"
-            android:hint="@string/profile_name"
-            android:singleLine="true"/>
+        android:layout_height="fill_parent"
+        android:paddingBottom="@dimen/content_vert_padding"
+        android:paddingLeft="@dimen/content_horz_padding"
+        android:paddingRight="@dimen/content_horz_padding"
+        android:paddingTop="@dimen/content_vert_padding">
 
         <LinearLayout
-            android:id="@+id/LinearLayoutIsSimpleRemote"
             android:layout_width="fill_parent"
             android:layout_height="wrap_content"
-            android:layout_marginBottom="8dp"
-            android:orientation="horizontal">
+            android:orientation="vertical">
 
-            <CheckBox
-                android:id="@+id/CheckBoxSimpleRemote"
-                android:layout_width="wrap_content"
-                android:layout_height="wrap_content"
-                android:text="@string/simple_remote"/>
-        </LinearLayout>
+            <android.support.design.widget.TextInputLayout
+                android:layout_width="fill_parent"
+                android:layout_height="wrap_content">
 
-        <TextView
-            style="@style/SimpleHeader"
-            android:layout_width="fill_parent"
-            android:layout_height="wrap_content"
-            android:layout_marginBottom="4dp"
-            android:text="@string/connection"/>
+                <EditText
+                    android:id="@+id/EditTextProfile"
+                    android:layout_width="fill_parent"
+                    android:layout_height="wrap_content"
+                    android:hint="@string/profile_name"
+                    android:singleLine="true"/>
+            </android.support.design.widget.TextInputLayout>
 
-        <LinearLayout
-            android:layout_width="fill_parent"
-            android:layout_height="wrap_content"
-            android:layout_marginBottom="4dp"
-            android:orientation="horizontal">
 
-            <EditText
-                android:id="@+id/EditTextHost"
-                android:layout_width="match_parent"
+            <LinearLayout
+                android:layout_width="fill_parent"
                 android:layout_height="wrap_content"
-                android:layout_marginRight="80dp"
-                android:hint="@string/host_long"
-                android:singleLine="true"/>
+                android:orientation="horizontal">
 
-            <EditText
-                android:id="@+id/EditTextPort"
-                android:layout_width="match_parent"
-                android:layout_height="wrap_content"
-                android:layout_marginLeft="-80dp"
-                android:hint="@string/port"
-                android:inputType="number"
-                android:maxLength="5"
-                android:singleLine="true"/>
-        </LinearLayout>
+                <android.support.design.widget.TextInputLayout
+                    android:layout_width="fill_parent"
+                    android:layout_height="wrap_content"
+                    android:layout_marginRight="80dp">
 
-        <LinearLayout
-            android:layout_width="fill_parent"
-            android:layout_height="wrap_content"
-            android:layout_marginBottom="8dp"
-            android:orientation="horizontal">
+                    <EditText
+                        android:id="@+id/EditTextHost"
+                        android:layout_width="match_parent"
+                        android:layout_height="wrap_content"
 
-            <CheckBox
-                android:id="@+id/CheckBoxSsl"
-                android:layout_width="wrap_content"
-                android:layout_height="wrap_content"
-                android:text="@string/ssl_enabled"/>
-        </LinearLayout>
+                        android:hint="@string/host_long"
+                        android:singleLine="true"/>
+                </android.support.design.widget.TextInputLayout>
 
-        <TextView
-            style="@style/SimpleHeader"
-            android:layout_width="fill_parent"
-            android:layout_height="wrap_content"
-            android:layout_marginBottom="4dp"
-            android:text="@string/login"/>
+                <android.support.design.widget.TextInputLayout
+                    android:layout_width="fill_parent"
+                    android:layout_height="wrap_content"
+                    android:layout_marginLeft="-80dp">
 
-        <LinearLayout
-            android:id="@+id/LinearLayoutIsLogin"
-            android:layout_width="fill_parent"
-            android:layout_height="wrap_content"
-            android:layout_marginBottom="4dp"
-            android:orientation="horizontal">
+                    <EditText
+                        android:id="@+id/EditTextPort"
+                        android:layout_width="match_parent"
+                        android:layout_height="wrap_content"
 
-            <CheckBox
-                android:id="@+id/CheckBoxLogin"
-                android:layout_width="wrap_content"
+                        android:hint="@string/port"
+                        android:inputType="number"
+                        android:maxLength="5"
+                        android:singleLine="true"/>
+                </android.support.design.widget.TextInputLayout>
+            </LinearLayout>
+
+            <LinearLayout
+                android:layout_width="fill_parent"
                 android:layout_height="wrap_content"
-                android:text="@string/login_enabled"/>
-        </LinearLayout>
+                android:orientation="horizontal">
 
-        <TableLayout
-            android:id="@+id/LoginLayout"
-            android:layout_width="fill_parent"
-            android:layout_height="wrap_content"
-            android:layout_marginBottom="4dp"
-            android:paddingLeft="6dp">
+                <CheckBox
+                    android:id="@+id/CheckBoxSsl"
+                    android:layout_width="wrap_content"
+                    android:layout_height="wrap_content"
+                    android:text="@string/ssl_enabled"/>
+            </LinearLayout>
 
-            <TableRow
-                android:id="@+id/tableRow1"
-                android:layout_width="wrap_content"
-                android:layout_height="wrap_content">
 
-                <TextView
+            <LinearLayout
+                android:id="@+id/LinearLayoutIsLogin"
+                android:layout_width="fill_parent"
+                android:layout_height="wrap_content"
+                android:orientation="horizontal">
+
+                <CheckBox
+                    android:id="@+id/CheckBoxLogin"
                     android:layout_width="wrap_content"
                     android:layout_height="wrap_content"
-                    android:layout_marginRight="8dp"
-                    android:text="@string/user"
-                    android:textAppearance="@android:style/TextAppearance.Medium"/>
+                    android:text="@string/login_enabled"/>
+            </LinearLayout>
 
-                <EditText
-                    android:id="@+id/EditTextUser"
-                    android:layout_width="fill_parent"
-                    android:layout_height="wrap_content"
-                    android:hint="@string/user"
-                    android:singleLine="true"/>
-            </TableRow>
 
-            <TableRow
-                android:id="@+id/tableRow2"
+            <LinearLayout
+                android:id="@+id/LoginLayout"
+                android:layout_width="fill_parent"
+                android:layout_height="wrap_content"
+                android:orientation="vertical">
+
+                <android.support.design.widget.TextInputLayout
+                    android:layout_width="200dp"
+                    android:layout_height="wrap_content">
+
+                    <EditText
+                        android:id="@+id/EditTextUser"
+                        android:layout_width="fill_parent"
+                        android:layout_height="wrap_content"
+                        android:hint="@string/user"
+                        android:singleLine="true"/>
+                </android.support.design.widget.TextInputLayout>
+
+                <android.support.design.widget.TextInputLayout
+                    android:layout_width="200dp"
+                    android:layout_height="wrap_content">
+
+                    <EditText
+                        android:id="@+id/EditTextPass"
+                        android:layout_width="fill_parent"
+                        android:layout_height="wrap_content"
+                        android:ems="10"
+                        android:hint="@string/pass"
+                        android:inputType="textPassword"
+                        android:singleLine="true"/>
+                </android.support.design.widget.TextInputLayout>
+            </LinearLayout>
+
+            <CheckBox
+                android:id="@+id/CheckBoxSimpleRemote"
                 android:layout_width="wrap_content"
-                android:layout_height="wrap_content">
+                android:layout_height="wrap_content"
+                android:layout_weight="1"
+                android:text="@string/simple_remote"/>
 
-                <TextView
-                    android:layout_width="wrap_content"
-                    android:layout_height="wrap_content"
-                    android:layout_marginRight="8dp"
-                    android:text="@string/pass"
-                    android:textAppearance="@android:style/TextAppearance.Medium"/>
+            <android.support.design.widget.TextInputLayout
+                android:layout_width="fill_parent"
+                android:layout_height="wrap_content">
 
                 <EditText
-                    android:id="@+id/EditTextPass"
+                    android:id="@+id/EditTextStreamHost"
                     android:layout_width="fill_parent"
                     android:layout_height="wrap_content"
-                    android:ems="10"
-                    android:hint="@string/pass"
-                    android:inputType="textPassword"
+                    android:hint="@string/stream_host_long"
                     android:singleLine="true"/>
-            </TableRow>
-        </TableLayout>
-
-        <TextView
-            style="@style/SimpleHeader"
-            android:layout_width="fill_parent"
-            android:layout_height="wrap_content"
-            android:layout_marginBottom="4dp"
-            android:layout_marginTop="4dp"
-            android:text="@string/stream"/>
-
-        <EditText
-            android:id="@+id/EditTextStreamHost"
-            android:layout_width="fill_parent"
-            android:layout_height="wrap_content"
-            android:layout_marginBottom="8dp"
-            android:hint="@string/stream_host_long"
-            android:singleLine="true"/>
+            </android.support.design.widget.TextInputLayout>
 
-        <LinearLayout
-            android:id="@+id/linearLayout4"
-            android:layout_width="fill_parent"
-            android:layout_height="wrap_content"
-            android:baselineAligned="false">
+            <CheckBox
+                android:id="@+id/CheckBoxEncoder"
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content"
+                android:layout_weight="1"
+                android:text="@string/use_encoder"/>
 
             <LinearLayout
+                android:id="@+id/linearLayoutEncoder"
                 android:layout_width="fill_parent"
                 android:layout_height="wrap_content"
-                android:layout_weight="1"
-                android:orientation="vertical"
-                android:paddingLeft="6dp"
-                android:paddingRight="6dp">
+                android:orientation="vertical">
 
-                <TextView
-                    style="@style/SimpleHeader"
+                <LinearLayout
+                    android:id="@+id/linearLayoutBitrates"
                     android:layout_width="fill_parent"
                     android:layout_height="wrap_content"
-                    android:layout_marginBottom="4dp"
-                    android:text="@string/live"/>
+                    android:orientation="horizontal">
+
+                    <android.support.design.widget.TextInputLayout
+                        android:layout_width="fill_parent"
+                        android:layout_height="wrap_content"
+                        android:layout_weight="1">
+
+                        <EditText
+                            android:id="@+id/EditTextEncoderPath"
+                            android:layout_width="fill_parent"
+                            android:layout_height="wrap_content"
+                            android:hint="@string/encoder_path"
+                            android:singleLine="true"
+                            android:text="/stream"/>
+                    </android.support.design.widget.TextInputLayout>
+
+                    <android.support.design.widget.TextInputLayout
+                        android:layout_width="fill_parent"
+                        android:layout_height="wrap_content"
+                        android:layout_weight="1">
+
+                        <EditText
+                            android:id="@+id/EditTextEncoderPort"
+                            android:layout_width="fill_parent"
+                            android:layout_height="wrap_content"
+                            android:hint="@string/encoder_port"
+                            android:inputType="number"
+                            android:maxLength="5"
+                            android:singleLine="true"
+                            android:text="554"/>
+                    </android.support.design.widget.TextInputLayout>
+                </LinearLayout>
 
                 <LinearLayout
-                    android:layout_width="wrap_content"
+                    android:id="@+id/linearLayoutBitrates"
+                    android:layout_width="fill_parent"
                     android:layout_height="wrap_content"
                     android:orientation="horizontal">
 
-                    <TextView
-                        android:layout_width="wrap_content"
+                    <android.support.design.widget.TextInputLayout
+                        android:layout_width="fill_parent"
                         android:layout_height="wrap_content"
-                        android:layout_marginLeft="3dp"
-                        android:layout_marginRight="8dp"
-                        android:text="@string/port"
-                        android:textAppearance="?android:attr/textAppearanceMedium"/>
-
-                    <EditText
-                        android:id="@+id/EditTextStreamPort"
-                        android:layout_width="80dp"
+                        android:layout_weight="1">
+
+                        <EditText
+                            android:id="@+id/EditTextVideoBitrate"
+                            android:layout_width="fill_parent"
+                            android:layout_height="wrap_content"
+                            android:hint="@string/video_bitrate"
+                            android:inputType="number"
+                            android:maxLength="4"
+                            android:singleLine="true"
+                            android:text="0"/>
+                    </android.support.design.widget.TextInputLayout>
+
+                    <android.support.design.widget.TextInputLayout
+                        android:layout_width="fill_parent"
                         android:layout_height="wrap_content"
-                        android:hint="@string/port"
-                        android:inputType="number"
-                        android:maxLength="5"
-                        android:singleLine="true">
-                    </EditText>
+                        android:layout_weight="1">
+
+                        <EditText
+                            android:id="@+id/EditTextAudioBitrate"
+                            android:layout_width="fill_parent"
+                            android:layout_height="wrap_content"
+                            android:hint="@string/audio_bitrate"
+                            android:inputType="number"
+                            android:maxLength="3"
+                            android:singleLine="true"
+
+                            android:text="0"/>
+                    </android.support.design.widget.TextInputLayout>
                 </LinearLayout>
-
-                <CheckBox
-                    android:id="@+id/CheckBoxLoginStream"
-                    android:layout_width="wrap_content"
-                    android:layout_height="wrap_content"
-                    android:layout_marginBottom="8dp"
-                    android:text="@string/login"/>
             </LinearLayout>
 
             <LinearLayout
+                android:id="@+id/linearLayoutStream"
                 android:layout_width="fill_parent"
                 android:layout_height="wrap_content"
-                android:layout_weight="1"
-                android:orientation="vertical"
-                android:paddingLeft="6dp"
-                android:paddingRight="6dp">
+                android:baselineAligned="false">
 
-                <TextView
-                    style="@style/SimpleHeader"
+                <LinearLayout
                     android:layout_width="fill_parent"
                     android:layout_height="wrap_content"
-                    android:layout_marginBottom="4dp"
-                    android:text="@string/movies"/>
+                    android:layout_weight="1"
+                    android:orientation="vertical"
+                    android:paddingLeft="6dp"
+                    android:paddingRight="6dp">
+
+                    <TextView
+                        style="@style/SimpleHeader"
+                        android:layout_width="fill_parent"
+                        android:layout_height="wrap_content"
+                        android:text="@string/live"/>
+
+                    <LinearLayout
+                        android:layout_width="wrap_content"
+                        android:layout_height="wrap_content"
+                        android:orientation="horizontal">
+
+
+                        <android.support.design.widget.TextInputLayout
+                            android:layout_width="fill_parent"
+                            android:layout_height="wrap_content">
+
+                            <EditText
+                                android:id="@+id/EditTextStreamPort"
+                                android:layout_width="80dp"
+                                android:layout_height="wrap_content"
+                                android:hint="@string/port"
+                                android:inputType="number"
+                                android:maxLength="5"
+                                android:singleLine="true">
+                            </EditText>
+                        </android.support.design.widget.TextInputLayout>
+                    </LinearLayout>
+
+                    <CheckBox
+                        android:id="@+id/CheckBoxLoginStream"
+                        android:layout_width="wrap_content"
+                        android:layout_height="wrap_content"
+                        android:text="@string/login"/>
+                </LinearLayout>
 
                 <LinearLayout
-                    android:layout_width="wrap_content"
+                    android:layout_width="fill_parent"
                     android:layout_height="wrap_content"
-                    android:orientation="horizontal">
+                    android:layout_weight="1"
+                    android:orientation="vertical"
+                    android:paddingLeft="6dp"
+                    android:paddingRight="6dp">
 
                     <TextView
+                        style="@style/SimpleHeader"
+                        android:layout_width="fill_parent"
+                        android:layout_height="wrap_content"
+                        android:text="@string/movies"/>
+
+                    <LinearLayout
                         android:layout_width="wrap_content"
                         android:layout_height="wrap_content"
-                        android:layout_marginLeft="3dp"
-                        android:layout_marginRight="8dp"
-                        android:text="@string/port"
-                        android:textAppearance="?android:attr/textAppearanceMedium"/>
+                        android:orientation="horizontal">
+
+                        <android.support.design.widget.TextInputLayout
+                            android:layout_width="fill_parent"
+                            android:layout_height="wrap_content">
+
+                            <EditText
+                                android:id="@+id/EditTextFilePort"
+                                android:layout_width="80dp"
+                                android:layout_height="wrap_content"
+                                android:hint="@string/port"
+                                android:inputType="number"
+                                android:maxLength="5"
+                                android:singleLine="true">
+                            </EditText>
+                        </android.support.design.widget.TextInputLayout>
+                    </LinearLayout>
+
+                    <CheckBox
+                        android:id="@+id/CheckBoxLoginFileStream"
+                        android:layout_width="wrap_content"
+                        android:layout_height="wrap_content"
+                        android:text="@string/login"/>
 
-                    <EditText
-                        android:id="@+id/EditTextFilePort"
-                        android:layout_width="80dp"
+                    <CheckBox
+                        android:id="@+id/CheckBoxSslFileStream"
+                        android:layout_width="wrap_content"
                         android:layout_height="wrap_content"
-                        android:hint="@string/port"
-                        android:inputType="number"
-                        android:maxLength="5"
-                        android:singleLine="true">
-                    </EditText>
+                        android:text="@string/ssl_enabled"/>
                 </LinearLayout>
-
-                <CheckBox
-                    android:id="@+id/CheckBoxLoginFileStream"
-                    android:layout_width="wrap_content"
-                    android:layout_height="wrap_content"
-                    android:text="@string/login"/>
-
-                <CheckBox
-                    android:id="@+id/CheckBoxSslFileStream"
-                    android:layout_width="wrap_content"
-                    android:layout_height="wrap_content"
-                    android:text="@string/ssl_enabled"/>
             </LinearLayout>
         </LinearLayout>
-    </LinearLayout>
 
-</android.support.v4.widget.NestedScrollView>
+    </android.support.v4.widget.NestedScrollView>
+
+    <com.melnykov.fab.FloatingActionButton
+        android:id="@+id/fab_save"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_alignParentBottom="true"
+        android:layout_alignParentEnd="true"
+        android:layout_alignParentRight="true"
+        android:layout_marginBottom="@dimen/fab_margin_bottom"
+        android:layout_marginRight="@dimen/fab_margin_right"
+        android:contentDescription="@string/reload"
+        android:src="@drawable/ic_action_save"/>
+</RelativeLayout>
\ No newline at end of file
diff --git a/app/res/values/strings.xml b/app/res/values/strings.xml
index abce9e8..4dac7d3 100644
--- a/app/res/values/strings.xml
+++ b/app/res/values/strings.xml
@@ -314,6 +314,11 @@
     <string name="privacy_statement_title">Privacy Statement</string>
     <string name="privacy">Privacy</string>
     <string name="next_event">Next Event</string>
+    <string name="use_encoder">Use encoder for streaming</string>
+    <string name="encoder_path">Encoder stream path</string>
+    <string name="audio_bitrate">Audio Bitrate</string>
+    <string name="video_bitrate">Video Bitrate</string>
+    <string name="encoder_port">Encoder Port</string>
     <string name="settings_and_about"><![CDATA[Settings & About]]></string>
     <string name="tools">Tools</string>
     <string name="control">Control</string>
diff --git a/app/src/net/reichholf/dreamdroid/Profile.java b/app/src/net/reichholf/dreamdroid/Profile.java
index ffa866d..ec6be13 100644
--- a/app/src/net/reichholf/dreamdroid/Profile.java
+++ b/app/src/net/reichholf/dreamdroid/Profile.java
@@ -22,17 +22,26 @@ public class Profile implements Serializable {
 	private String mName;
 	private String mHost;
 	private String mStreamHost;
+	private String mEncoderPath;
 	private String mUser;
 	private String mPass;
+	private String mEncoderUser;
+	private String mEncoderPass;
 	private boolean mLogin;
 	private boolean mSsl;
 	private boolean mStreamLogin;
 	private boolean mFileLogin;
+	private boolean mEncoderLogin;
+	private boolean mEncoderStream;
 	private boolean mFileSsl;
 	private boolean mSimpleRemote;
 	private int mPort;
 	private int mStreamPort;
 	private int mFilePort;
+	private int mEncoderPort;
+	private int mEncoderAudioBitrate;
+	private int mEncoderVideoBitrate;
+
 	private String mDefaultRef;
 	private String mDefaultRefName;
 	private String mDefaultRef2;
@@ -61,6 +70,50 @@ public class Profile implements Serializable {
 		setSimpleRemote(simpleRemote);
 		setDefaultRefValues(defaultRef, defaultRefName);
 		setDefaultRef2Values(defaultRef2, defaultRef2Name);
+
+		//Encoder
+		setEncoderPort(554);
+		setEncoderPath("stream");
+		setEncoderAudioBitrate(128);
+		setEncoderVideoBitrate(3000);
+		setEncoderLogin(false);
+		setEncoderUser("");
+		setEncoderPass("");
+
+	}
+
+	public Profile(int id, String name, String host, String streamHost, int port, int streamPort, int filePort, boolean login,
+				   String user, String pass, boolean ssl, boolean streamLogin, boolean fileLogin, boolean fileSsl,
+				   boolean simpleRemote, String defaultRef, String defaultRefName, String defaultRef2, String defaultRef2Name, boolean encoderStream,
+					int encoderPort, String encoderPath, int encoderVideoBitrate, int encoderAudioBitrate, boolean encoderLogin, String encoderUser, String encoderPass) {
+
+		mId = id;
+		mSessionId = null;
+		setName(name);
+		setHost(host);
+		setStreamHost(streamHost);
+		setPort(port);
+		setStreamPort(streamPort);
+		setFilePort(filePort);
+		setLogin(login);
+		setStreamLogin(streamLogin);
+		setFileLogin(fileLogin);
+		setFileSsl(fileSsl);
+		setUser(user);
+		setPass(pass);
+		setSsl(ssl);
+		setSimpleRemote(simpleRemote);
+		setDefaultRefValues(defaultRef, defaultRefName);
+		setDefaultRef2Values(defaultRef2, defaultRef2Name);
+		//Encoder
+		setEncoderStream(encoderStream);
+		setEncoderPort(encoderPort);
+		setEncoderPath(encoderPath);
+		setEncoderAudioBitrate(encoderAudioBitrate);
+		setEncoderVideoBitrate(encoderVideoBitrate);
+		setEncoderLogin(encoderLogin);
+		setEncoderUser(encoderUser);
+		setEncoderPass(encoderPass);
 	}
 
 	public void setPort(int port) {
@@ -120,6 +173,14 @@ public class Profile implements Serializable {
 		return mStreamHost;
 	}
 
+	public void setEncoderPath(String encoderPath) {
+		mEncoderPath = encoderPath;
+	}
+
+	public String getEncoderPath() {
+		return mEncoderPath;
+	}
+
 	public String getUser() {
 		return mUser;
 	}
@@ -148,6 +209,30 @@ public class Profile implements Serializable {
 		mLogin = login;
 	}
 
+	public String getEncoderUser() {
+		return mEncoderUser;
+	}
+
+	public void setEncoderUser(String user) {
+		mEncoderUser = user;
+	}
+
+	public String getEncoderPass() {
+		return mEncoderPass;
+	}
+
+	public void setEncoderPass(String pass) {
+		mEncoderPass = pass;
+	}
+
+	public boolean isEncoderLogin() {
+		return mEncoderLogin;
+	}
+
+	public void setEncoderLogin(boolean isLogin) {
+		mEncoderLogin = isLogin;
+	}
+
 	public boolean isSsl() {
 		return mSsl;
 	}
@@ -218,10 +303,44 @@ public class Profile implements Serializable {
 		}
 	}
 
+
+	public boolean isEncoderStream() {
+		return mEncoderStream;
+	}
+
+	public void setEncoderStream(boolean encoderStream) {
+		mEncoderStream = encoderStream;
+	}
+
 	public String getStreamPortString() {
 		return String.valueOf(mStreamPort);
 	}
 
+	public int getEncoderPort() {
+		return mEncoderPort;
+	}
+
+	public void setEncoderPort(int port) {
+		mEncoderPort = port;
+	}
+
+	public int getEncoderVideoBitrate() {
+		return mEncoderVideoBitrate;
+	}
+
+	public void setEncoderVideoBitrate(int bitrate) {
+		mEncoderVideoBitrate = bitrate;
+	}
+
+	public int getEncoderAudioBitrate() {
+		return mEncoderAudioBitrate;
+	}
+
+	public void setEncoderAudioBitrate(int bitrate) {
+		mEncoderAudioBitrate = bitrate;
+	}
+
+
 	public int getFilePort() {
 		return mFilePort;
 	}
@@ -289,21 +408,36 @@ public class Profile implements Serializable {
 		mStreamLogin = streamLogin;
 	}
 
-	public void setSessionId(String sessionId){
+	public void setSessionId(String sessionId) {
 		mSessionId = sessionId;
 	}
 
-	public String getSessionId(){
+	public String getSessionId() {
 		return mSessionId;
 	}
 
 	public boolean equals(Profile p) {
-		return getHost().equals(p.getHost()) && getStreamHost().equals(p.getStreamHost())
-				&& getUser().equals(p.getUser()) && getPass().equals(p.getPass()) && isLogin() == p.isLogin()
-				&& isSsl() == p.isSsl() && isSimpleRemote() == p.isSimpleRemote() && getId() == p.getId()
-				&& getPort() == p.getPort() && getStreamPort() == p.getStreamPort() && getFilePort() == p.getFilePort()
-				&& isStreamLogin() == p.isStreamLogin() && isFileSsl() == p.isFileSsl()
-				&& isFileLogin() == p.isFileLogin();
+		return getHost().equals(p.getHost())
+			&& getStreamHost().equals(p.getStreamHost())
+			&& getUser().equals(p.getUser())
+			&& getPass().equals(p.getPass())
+			&& isLogin() == p.isLogin()
+			&& isSsl() == p.isSsl()
+			&& isSimpleRemote() == p.isSimpleRemote()
+			&& getId() == p.getId()
+			&& getPort() == p.getPort()
+			&& getStreamPort() == p.getStreamPort()
+			&& getFilePort() == p.getFilePort()
+			&& isStreamLogin() == p.isStreamLogin()
+			&& isFileSsl() == p.isFileSsl()
+			&& isFileLogin() == p.isFileLogin()
+			&& getEncoderPort() == p.getEncoderPort()
+			&& isEncoderStream() == p.isEncoderStream()
+			&& isEncoderLogin() == p.isEncoderLogin()
+			&& getEncoderVideoBitrate() == p.getEncoderVideoBitrate()
+			&& getEncoderAudioBitrate() == p.getEncoderAudioBitrate()
+			&& getEncoderPath().equals(p.getEncoderPath())
+			&& getEncoderUser().equals(p.getEncoderUser())
+			&& getEncoderPass().equals(p.getEncoderPass());
 	}
-
 }
diff --git a/app/src/net/reichholf/dreamdroid/fragment/ProfileEditFragment.java b/app/src/net/reichholf/dreamdroid/fragment/ProfileEditFragment.java
index 52dace3..d1dfd5c 100644
--- a/app/src/net/reichholf/dreamdroid/fragment/ProfileEditFragment.java
+++ b/app/src/net/reichholf/dreamdroid/fragment/ProfileEditFragment.java
@@ -39,19 +39,26 @@ public class ProfileEditFragment extends DreamDroidFragment {
 
 	private EditText mProfile;
 	private EditText mHost;
+	private EditText mUser;
+	private EditText mPass;
 	private EditText mStreamHost;
 	private EditText mPort;
 	private EditText mStreamPort;
 	private EditText mFilePort;
+	private EditText mEncoderPath;
+	private EditText mEncoderVideoBitrate;
+	private EditText mEncoderAudioBitrate;
+	private EditText mEncoderPort;
 	private CheckBox mSsl;
 	private CheckBox mLogin;
+	private CheckBox mEncoderStream;
 	private CheckBox mStreamLogin;
 	private CheckBox mFileSsl;
 	private CheckBox mFileLogin;
-	private EditText mUser;
-	private EditText mPass;
 	private CheckBox mSimpleRemote;
+	private LinearLayout mLayoutEncoder;
 	private LinearLayout mLayoutLogin;
+	private LinearLayout mLayoutStream;
 
 	@Override
 	public void onCreate(Bundle savedInstanceState) {
@@ -80,8 +87,19 @@ public class ProfileEditFragment extends DreamDroidFragment {
 		mFileSsl = (CheckBox) view.findViewById(R.id.CheckBoxSslFileStream);
 		mFileLogin = (CheckBox) view.findViewById(R.id.CheckBoxLoginFileStream);
 
+		//Encoder
+		mEncoderStream = (CheckBox) view.findViewById(R.id.CheckBoxEncoder);
+		mEncoderPath = (EditText) view.findViewById(R.id.EditTextEncoderPath);
+		mEncoderPort = (EditText) view.findViewById(R.id.EditTextEncoderPort);
+		mEncoderVideoBitrate = (EditText) view.findViewById(R.id.EditTextVideoBitrate);
+		mEncoderAudioBitrate = (EditText) view.findViewById(R.id.EditTextAudioBitrate);
+
+		mLayoutEncoder = (LinearLayout) view.findViewById(R.id.linearLayoutEncoder);
+		mLayoutStream = (LinearLayout) view.findViewById(R.id.linearLayoutStream);
 		mLayoutLogin = (LinearLayout) view.findViewById(R.id.LoginLayout);
 
+
+
 		if (Intent.ACTION_EDIT.equals(getArguments().getString("action"))) {
 			mCurrentProfile = (Profile) getArguments().getSerializable("profile");
 			if (mCurrentProfile == null)
@@ -89,7 +107,24 @@ public class ProfileEditFragment extends DreamDroidFragment {
 			assignProfile();
 		}
 		onIsLoginChanged(mLogin.isChecked());
+		onIsEncoderStreamChanged(mEncoderStream.isChecked());
 
+		registerListeners();
+		registerFab(R.id.fab_main, view, R.string.save, R.drawable.ic_action_save, new View.OnClickListener() {
+			@Override
+			public void onClick(View v) {
+				save();
+			}
+		});
+		return view;
+	}
+
+	@Override
+	public void createOptionsMenu(Menu menu, MenuInflater inflater) {
+		inflater.inflate(R.menu.cancel, menu);
+	}
+
+	private void registerListeners(){
 		mLogin.setOnCheckedChangeListener(new OnCheckedChangeListener() {
 			@Override
 			public void onCheckedChanged(CompoundButton checkbox, boolean checked) {
@@ -103,18 +138,13 @@ public class ProfileEditFragment extends DreamDroidFragment {
 				onSslChanged(checked);
 			}
 		});
-		registerFab(R.id.fab_main, view, R.string.save, R.drawable.ic_action_save, new View.OnClickListener() {
+		mEncoderStream.setOnCheckedChangeListener(new OnCheckedChangeListener() {
 			@Override
-			public void onClick(View v) {
-				save();
+			public void onCheckedChanged(CompoundButton buttonView, boolean checked) {
+				onIsEncoderStreamChanged(checked);
 			}
 		});
-		return view;
-	}
 
-	@Override
-	public void createOptionsMenu(Menu menu, MenuInflater inflater) {
-		inflater.inflate(R.menu.cancel, menu);
 	}
 
 	public boolean onOptionsItemSelected(MenuItem item) {
@@ -150,6 +180,16 @@ public class ProfileEditFragment extends DreamDroidFragment {
 		}
 	}
 
+	private void onIsEncoderStreamChanged(boolean checked) {
+		if (checked) {
+			mLayoutEncoder.setVisibility(View.VISIBLE);
+			mLayoutStream.setVisibility(View.GONE);
+		} else {
+			mLayoutEncoder.setVisibility(View.GONE);
+			mLayoutStream.setVisibility(View.VISIBLE);
+		}
+	}
+
 	/**
 	 * Assign all values of <code>mProfile</code> to the GUI-Components
 	 */
@@ -168,6 +208,11 @@ public class ProfileEditFragment extends DreamDroidFragment {
 		mFileLogin.setChecked(mCurrentProfile.isFileLogin());
 		mFileSsl.setChecked(mCurrentProfile.isFileSsl());
 		mSimpleRemote.setChecked(mCurrentProfile.isSimpleRemote());
+
+		mEncoderPath.setText(mCurrentProfile.getEncoderPath());
+		mEncoderPort.setText(String.valueOf(mCurrentProfile.getEncoderPort()));
+		mEncoderVideoBitrate.setText(String.valueOf(mCurrentProfile.getEncoderVideoBitrate()));
+		mEncoderAudioBitrate.setText(String.valueOf(mCurrentProfile.getEncoderAudioBitrate()));
 	}
 
 	/**
@@ -187,6 +232,12 @@ public class ProfileEditFragment extends DreamDroidFragment {
 		mCurrentProfile.setUser(mUser.getText().toString());
 		mCurrentProfile.setPass(mPass.getText().toString());
 		mCurrentProfile.setSimpleRemote(mSimpleRemote.isChecked());
+		//Encoder
+		mCurrentProfile.setEncoderAudioBitrate(Integer.valueOf(mEncoderAudioBitrate.getText().toString()));
+		mCurrentProfile.setEncoderVideoBitrate(Integer.valueOf(mEncoderVideoBitrate.getText().toString()));
+		mCurrentProfile.setEncoderPort(Integer.valueOf(mEncoderPort.getText().toString()));
+
+
 		DatabaseHelper dbh = DatabaseHelper.getInstance(getAppCompatActivity());
 		if (mCurrentProfile.getId() > 0) {
 			if (mCurrentProfile.getHost() == null || "".equals(mCurrentProfile.getHost())) {
diff --git a/app/src/net/reichholf/dreamdroid/helpers/SimpleHttpClient.java b/app/src/net/reichholf/dreamdroid/helpers/SimpleHttpClient.java
index 88d3194..81e4dec 100644
--- a/app/src/net/reichholf/dreamdroid/helpers/SimpleHttpClient.java
+++ b/app/src/net/reichholf/dreamdroid/helpers/SimpleHttpClient.java
@@ -110,6 +110,33 @@ public class SimpleHttpClient {
 	 * @param title
 	 * @return
 	 */
+	public String buildEncoderStreamUrl(String ref, String title) {
+		try {
+			ref = URLEncoder.encode(ref, "utf-8").replace("+", "%20");
+		} catch (UnsupportedEncodingException e) {
+		}
+		String streamLoginString = "";
+		if (mProfile.isEncoderLogin())
+			streamLoginString = mProfile.getEncoderUser() + ":" + mProfile.getEncoderPass() + "@";
+
+		String url = String.format(
+				"rtsp://%s%s:%s/%s?ref=%s&video_bitrate=%s&audio_bitrate=%s",
+				streamLoginString,
+				mProfile.getStreamHost(),
+				mProfile.getEncoderPort(),
+				mProfile.getEncoderPath(),
+				ref,
+				mProfile.getEncoderVideoBitrate(),
+				mProfile.getEncoderAudioBitrate()
+		);
+		return url;
+	}
+
+	/**
+	 * @param ref
+	 * @param title
+	 * @return
+	 */
 	public String buildServiceStreamUrl(String ref, String title) {
 		try {
 			ref = URLEncoder.encode(ref, "utf-8").replace("+", "%20");
diff --git a/app/src/net/reichholf/dreamdroid/intents/IntentFactory.java b/app/src/net/reichholf/dreamdroid/intents/IntentFactory.java
index fdf5fce..0a078db 100644
--- a/app/src/net/reichholf/dreamdroid/intents/IntentFactory.java
+++ b/app/src/net/reichholf/dreamdroid/intents/IntentFactory.java
@@ -49,10 +49,14 @@ public class IntentFactory {
 
 	public static Intent getStreamServiceIntent(String ref, String title) {
 		Intent intent = new Intent(Intent.ACTION_VIEW);
-		String uriString = SimpleHttpClient.getInstance().buildServiceStreamUrl(ref, title);
+		String uriString = SimpleHttpClient.getInstance().buildEncoderStreamUrl(ref, title);
 		Log.i(DreamDroid.LOG_TAG, "Service-Streaming URL set to '" + uriString + "'");
 
-		intent.setDataAndType(Uri.parse(uriString), "video/*");
+		Uri uri = Uri.parse(uriString);
+		if(uri.getScheme() == "http")
+			intent.setDataAndType(uri, "video/*");
+		else
+			intent.setData(uri);
 		intent.putExtra("title", title);
 		return intent;
 	}
-- 
1.9.1

